x-pack/test/ingest_manager_api_integration/apis/epm/setup.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import expect from '@kbn/expect';
+import { FtrProviderContext } from '../../../api_integration/ftr_provider_context';
+import { skipIfNoDockerRegistry } from '../../helpers';
+import { GetInfoResponse, Installed } from '../../../../plugins/ingest_manager/common';
+
+export default function (providerContext: FtrProviderContext) {
+  const { getService } = providerContext;
+  const supertest = getService('supertest');
+  const log = getService('log');
+
+  describe('setup api', async () => {
```

Comment:
Can you add some tests that show the failure case(s)? I'm unclear what we intend to be returned from a `POST /setup` that has a failed upgrade at some point. Is it a 4xx or 5xx response? What's in the response body?

Reply:
Yeah so adding the failure case might be tricky. Any ideas on how to trigger a failure? This is what comes to mind:

1. Override the default packages variable used to define which packages are installed/updated during `/setup` so I can create a custom one and make it malformed. Is this possible with mocha?
2. Use an invalid registry address so kibana fails to download the packages. I believe this would require creating a new `x-pack/test` directory because the tests in this directory all use the docker container registry
3. Create a malformed `endpoint` package in this directory and make the version really big like 100000.0.0 so it's always the **latest**. The number needs to be large so it's unlikely to be superseded in production. The test would start failing as soon as we created a production package with version 100000.0.1 though 🤔

Reply:
If we can map out the desired behavior I am happy to help with testing. I did a few different types recently working on Registry and /setup changes.

If this is run on every setup I really want to understand/document what it should/does do.

Reply:
I'll have to do some digging to figure out what should be returned in each failure scenario. After thinking about this more, the code changes will currently swallow the error :( because of this:

```
if (isBulkInstallError(resp)) {
      throw new Error(resp.error);
    }
```

from here: https://github.com/elastic/kibana/pull/78081/files#diff-bae6a46773afa98cb0927abf59bdd003R93

That's probably not what we want but in the current state if an error occurs it will always 500 and the response body will look like this:

```
body: {
  message: "some error"
}
```
https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/server/errors/handlers.ts#L100

The success case will always return:
200
```
{
  isInitialized: true
}
```

For the error case we could probably create a new error class that is a child of `IngestManagerError` and modify this function: https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/server/errors/handlers.ts#L48

to extract out the passed in status code. That way we can preserve the status code and message that occurred when the upgrade/install failed for a package and return that as a response for `/setup`. That way the failure functionality will change.

Reply:
@skh @neptunian any ideas on how to force a failure scenario in `/setup`? Is there a way to mock out `DefaultPackages` so that I can craft a fake package that causes a failure when being installed?
https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/common/types/models/epm.ts#L265